File: src/productrank/routers/results.py
"""GET /v1/results — the latest saved evaluation results (for the Analytics page).

Reads results/eval_{split}.json produced by `productrank eval`. Returning the measured
table over an API keeps the frontend a thin renderer with no duplicated numbers.
"""

from __future__ import annotations

import json
from pathlib import Path

from fastapi import APIRouter, HTTPException

from productrank.config import DATASET_SPLIT
from productrank.schemas import Dataset

router = APIRouter(prefix="/v1", tags=["results"])

RESULTS_DIR = Path("results")


@router.get("/results")
def get_results(dataset: Dataset = Dataset.MSMARCO) -> dict:
    # dataset → split (msmarco→dev, fiqa→test). The dashboard toggles by dataset; this is
    # the single mapping point, so search and the dashboard always reflect the same corpus.
    split = DATASET_SPLIT[dataset.value]
    path = RESULTS_DIR / f"eval_{split}.json"
    if not path.exists():
        raise HTTPException(404, f"no results for dataset={dataset.value}; run `productrank eval`")
    data = json.loads(path.read_text())
    # Trim per-query payload — the dashboard only needs the aggregate table + timings.
    return {
        "split": data.get("split"),
        "num_queries": data.get("num_queries"),
        "top_k": data.get("top_k"),
        "aggregate": data.get("aggregate", {}),
        "wall_seconds": data.get("wall_seconds", {}),
    }

File: frontend/lib/results.ts
import type { Dataset, Variant } from "./api";
import { DEFAULT_DATASET } from "./api";

export interface ResultsResponse {
  split: string;
  num_queries: number;
  top_k: number;
  aggregate: Record<Variant, Record<string, number>>;
  wall_seconds: Record<string, number>;
}

const BASE = process.env.NEXT_PUBLIC_API_BASE || "/api";

// The backend maps dataset → split (msmarco→dev, fiqa→test), so the dashboard and live
// search always reflect the same corpus.
export async function getResults(dataset: Dataset = DEFAULT_DATASET): Promise<ResultsResponse> {
  const res = await fetch(`${BASE}/v1/results?dataset=${dataset}`, { cache: "no-store" });
  if (!res.ok) throw new Error(`/v1/results → ${res.status}`);
  return res.json();
}
